<a href="https://colab.research.google.com/github/Jojomaque/DataScience/blob/main/JojoSorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning 
---



In order to effectively analyse a dataset, often we need to prepare it first. 
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* Sort by a single column e.g.
  * `df.sort_values("Make") `
* Sort by multiple columns e.g. 
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make 
* Sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted* 

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score 
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [ ]:
import pandas as pd
df = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true')
df
happiness = df
#Display first 5 rows
happiness.head()
sort_by_happiness_rank = happiness.sort_values(by = "Happiness Score",ascending = True)
display(sort_by_happiness_rank)

Country  ... Dystopia Residual
157         Togo  ...           1.56726
156      Burundi  ...           1.83302
155        Syria  ...           0.32858
154        Benin  ...           1.63328
153       Rwanda  ...           0.67042
..           ...  ...               ...
4         Canada  ...           2.45176
3         Norway  ...           2.46531
2        Denmark  ...           2.49204
1        Iceland  ...           2.70201
0    Switzerland  ...           2.51738

[158 rows x 12 columns]

### Exercise 2 - sort by multiple columns, display the first 5 rows 
---

1. Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order 
2. Display the first 5 rows of sorted data 

**Test output**:  
Records 122, 127, 147, 100, 96

In [11]:
import pandas as pd
df = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true')
df
happiness_original = df
#Display first 5 rows
happiness_original.head()
happiness_original.sort_values(['Economy (GDP per Capita)','Health (Life Expectancy)'])[['Economy (GDP per Capita)','Health (Life Expectancy)']].head()

Economy (GDP per Capita)  Health (Life Expectancy)
119                   0.00000                   0.09806
156                   0.01530                   0.22396
130                   0.01604                   0.22562
143                   0.06940                   0.29707
115                   0.07120                   0.34201

### Exercise 3 - sorting in descending order 
---
 
Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
43, 7, 144, 0, 3 Country and Region columns


In [15]:
happiness_original.sort_values(['Freedom','Trust (Government Corruption)'],ascending = False)[['Country','Region']].tail()  #Please check my code as mu out put is different

Country                           Region
136                  Angola               Sub-Saharan Africa
117                   Sudan               Sub-Saharan Africa
95   Bosnia and Herzegovina       Central and Eastern Europe
101                  Greece                   Western Europe
111                    Iraq  Middle East and Northern Africa

# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values 

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values 
---

1. Read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [18]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv')
housing_london_original = df
housing_london_original.isnull().values.any()

True

In [20]:
#.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)
housing_london_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


### Exercise 5 - remove null values 
---

1. Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. Remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [22]:
#Remove rows with NA values for life_satisfaction
df.dropna(subset = ['life_satisfaction'])

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
1010  E12000009            south west  ...          NaN             0
1013  E92000001               england  ...   24172166.0             0
1017  N92000002      northern ireland  ...          NaN             0
1018  S92000003              scotland  ...          NaN             0
1019  W92000004                 wales  ...          NaN             0

[352 rows x 12 columns]

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1 

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.
